pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -0,0 +1,289 @@
+	clusterLister         v3.ClusterLister
+	backupClient          v3.EtcdBackupInterface
+	backupLister          v3.EtcdBackupLister
+	backupDriver          service.EngineService
+	secretsClient         v1.SecretInterface
+	KontainerDriverLister v3.KontainerDriverLister
+}
+
+func Register(ctx context.Context, management *config.ManagementContext) {
+	c := &Controller{
+		ctx:                   ctx,
+		clusterClient:         management.Management.Clusters(""),
+		clusterLister:         management.Management.Clusters("").Controller().Lister(),
+		backupClient:          management.Management.EtcdBackups(""),
+		backupLister:          management.Management.EtcdBackups("").Controller().Lister(),
+		backupDriver:          service.NewEngineService(clusterprovisioner.NewPersistentStore(management.Core.Namespaces(""), management.Core)),
+		secretsClient:         management.Core.Secrets(""),
+		KontainerDriverLister: management.Management.KontainerDrivers("").Controller().Lister(),
+	}
+
+	local := &rkedialerfactory.RKEDialerFactory{
+		Factory: management.Dialer,
+	}
+	docker := &rkedialerfactory.RKEDialerFactory{
+		Factory: management.Dialer,
+		Docker:  true,
+	}
+	driver := service.Drivers[service.RancherKubernetesEngineDriverName]
+	rkeDriver := driver.(*rke.Driver)
+	rkeDriver.DockerDialer = docker.Build
+	rkeDriver.LocalDialer = local.Build
+	rkeDriver.WrapTransportFactory = docker.WrapTransport
+
+	c.backupClient.AddLifecycle(ctx, "etcdbackup-controller", c)
+	go c.clusterBackupSync(ctx, clusterBackupCheckInterval)
+}
+
+func (c *Controller) Create(b *v3.EtcdBackup) (runtime.Object, error) {
+	cluster, err := c.clusterClient.Get(b.ClusterID, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig == nil {
+		return nil, fmt.Errorf("[etcd-backup] cluster doesn't have a backup config")
+	}
+
+	if !v3.BackupConditionCompleted.IsTrue(b) {
+		b.Filename = getBackupFilename(b.Name, cluster)
+		b.BackupConfig = *cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig
+		v3.BackupConditionCompleted.True(b)
+		b, err = c.backupClient.Update(b)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	kontainerDriver, err := c.KontainerDriverLister.Get("", service.RancherKubernetesEngineDriverName)
+	if err != nil {
+		return nil, err
+	}
+	bObj, err := v3.BackupConditionCompleted.Do(b, func() (runtime.Object, error) {
+		err = c.backupDriver.ETCDSave(context.Background(), cluster.Name, kontainerDriver, cluster.Spec, b.Name)
+		return b, err
+	})
+	if err != nil {
+		logrus.Errorf("[etcd-backup] failed to perform etcd backup: %v", err)
+	}
+	// need to update the object to keep the condition change.
+	return c.backupClient.Update(bObj.(*v3.EtcdBackup))
+}
+
+func (c *Controller) Remove(b *v3.EtcdBackup) (runtime.Object, error) {
+	logrus.Infof("[etcd-backup] Deleteing backup %s ", b.Name)
+	if err := c.deleteS3Snapshot(b); err != nil {
+		return nil, err
```

Comment:
I think you need to return non=nil object regardless. Why don't we just `return b, c.deleteS3Snapshot(b)`